hal/spi_api.h

```diff
@@ -19,220 +18,331 @@
- * @param[in]     mode  The SPI mode (clock polarity, phase, and shift direction)
- * @param[in]     slave Zero for master mode or non-zero for slave mode
+
+/**
+ * This enumerates the possible event types generated by the SPI ASYNC api.
  */
-void spi_format(spi_t *obj, int bits, int mode, int slave);
+typedef enum spi_event_type_t {
+    SPI_EVENT_TYPE_ON_DONE, /**< The operation has completed successfully. */
+    SPI_EVENT_TYPE_ON_ABORT, /**< The operation has been aborted. */
+    SPI_EVENT_TYPE_ON_ERROR /**< An error occured. */
+} spi_even_type_t;
+
 
-/** Set the SPI baud rate
+/**
+ * Signature for an SPI async completion event.
  *
- * Actual frequency may differ from the desired frequency due to available dividers and bus clock
- * Configures the SPI peripheral's baud rate
- * @param[in,out] obj The SPI object to configure
- * @param[in]     hz  The baud rate in Hz
+ * As this may be executed from an interrupt context it is highly adviced to restrict this callback
+ * to signaling completion to a thread.
  */
-void spi_frequency(spi_t *obj, int hz);
+typedef void (*spi_event_f)(void *context, spi_event_type_t evtype);
+
+#ifdef __cplusplus
+extern "C" {
+#endif
 
-/**@}*/
 /**
- * \defgroup SynchSPI Synchronous SPI Hardware Abstraction Layer
- * @{
- */
+ * Initializes a SPI instance.
 
-/** Write a byte out in master mode and receive a value
+ * A single SPI peripheral might be used by multiple spi_t as it only represents a communication
+ * channel towards a slave (or a master). Instances of spi_t may only differ by their SS pin.
  *
- * @param[in] obj   The SPI peripheral to use for sending
- * @param[in] value The value to send
- * @return Returns the value received during send
+ * @param[in,out] obj   This.
+ * @param[in]     init  Initialization parameters.
+ *
+ * @return SPI_RESULT_OK on success. See spi_result_t for more details about failures.
  */
-int  spi_master_write(spi_t *obj, int value);
+spi_result_t spi_init(spi_t *obj, spi_init_t *init);
 
-/** Write a block out in master mode and receive a value
+/**
+ * Processes a transfer blocking until completion.
+ * This function locks the peripheral and prevents any other transaction to occur until the
+ * operation is completed (including asynchronous transactions).
+ * It will return `true` on success or `false` if :
+ * - at least one of its parameters is undefined (NULL).
+ * - the peripheral is busy with another transaction (from this spi_t or any other sharing the same peripheral).
+ * - any error occurs during the transfer. Errors can be but are not limited to :
+ *   - conflicting master ;
+ *   - rx or tx buffer overflow ;
+ *   - rx or tx buffer underflow.
  *
- *  The total number of bytes sent and received will be the maximum of
- *  tx_length and rx_length. The bytes written will be padded with the
- *  value 0xff.
+ * It sends `tx_count` symbols from the buffer pointed by `tx` or a place holder is `tx` is NULL.
```

Comment:
`r a place holder is `tx` is NULL` -> `r a place holder if `tx` is NULL`

Also you are using `place holder` here and `fill symbol` before can we align please.

Reply:
good catch 👍